execution environments: add EE overview page (#60)
ansible/community-docs: execution_environments/execution_environments_overview.rst

.. _ee_overview:

***************************************
Ansible Execution Environments Overview
***************************************

.. contents::
   :local:

.. _terminology:

Terminology
===========

Notions used in this document:

* **ansible-core**: you install it using ``pip install ansible-core``; includes command-line tools such as ``ansible-playbook`` and ``ansible-galaxy``, the Ansible language and a set of `builtin modules and plugins <https://docs.ansible.com/ansible/latest/collections/ansible/builtin/index.html>`_.
* **Ansible collections**: a format in which Ansible content is distributed that can contain playbooks, roles, modules and plugins.
* **Ansible package**: you install it using ``pip install ansible`` or an OS distribution package manager; it provides ansible-core and a big set of Ansible collections in the *batteries included* manner.
* **Ansible**: in the context of this document, is the Ansible package or ansible-core plus a set of collections installed on the Ansible control node.
* **Ansible control node**: the machine from which you run Ansible.

> * **Containerization**: it's what you typically use ``podman`` or ``docker`` for.

Reviewer: ```suggestion
* **Container Runtime**: it is what you typically use ``podman`` or ``docker`` for running containers.
```

> * **Container Runtime**: it is what you typically use ``podman`` or ``docker`` for running containers.

Author: SGTM, I added a short and very general definition. The more details explanation will come right below in the next section.

.. _ee_rationale:

Why Execution Environments were introduced

> ==========================================

Author: changed across the docs, thanks

> The Ansible automation execution environments were introduced to resolve the following issues

Reviewer: ```suggestion
The Ansible Execution Environments were introduced to resolve the following issues
```

Dependencies
------------

Software applications typically have dependencies.
It can be software libraries, configuration files or other services, to name a few, and Ansible,
being an exceptional automation tool, is not an exception in terms of the mentioned.

Traditionally, application dependencies are installed by administrators on top of
an operating system using packaging management tools like RPM or Python-pip.

The major drawback of such approach is that an application might require versions
of dependencies different from those provided by default.

In case of Ansible, a typical installation consists of ansible-core and a set of Ansible collections.

At present, there are more than a hundred collections included in the Ansible package and
hundreds more are available on Ansible Galaxy and Automation Hub for manual installation.
Many of them have dependencies.

The Ansible collections can depend on the following pieces of software and their versions:

* ansible-core 
* Python
* Python packages
* system packages
* other Ansible collections

The dependencies have to be installed and sometimes can conflict with each other.

One way to **partly** resolve dependency issue is

> to use Python virtual environments on Ansible controle node.

Reviewer: ```suggestion
to use Python virtual environments on Ansible control node.
```

However, applied to Ansible, it has drawbacks and natural limitations.

Content separation
------------------

Typically users don't really need the whole Ansible package with that hundred collections.
In most cases, and it's considered a good practice, you need to install only ``ansible-core``
plus a set of collections for specific tasks, nothing more.

Let's say, there's an Ansible control node or a tool like Ansible AWX/Controller used by several users.

Each user works with a limited set of services and wants to use only ``ansible-core``
and a corresponding set of Ansible collections automating those services.

> Each user is not interested in having someone else's content the users don't need including dependencies

Reviewer: ```suggestion
Each user is not interested in having someone else's content the users do not need including dependencies
```

which can potentially brake their pipelines.
They might also want to preserve their installation clean from unnecessary packages
and files belonging their colleagues.

Simply put, users might want their content be separated.

Portability
-----------

An Ansible user writes content for Ansible locally and wants to leverage the container technology
to make your automation runtimes portable, shareable and easily deployable to testing and production environments.

What is Execution Environment
=============================

Ansible, as a software application, can run in a container, thus, it can benefit from containerization the same as most other applications.

The Ansible automation execution environment (hereinafter, execution environment or EE) is a container image serving as Ansible control node.

The EE image contains:

* ansible-core
* none or more Ansible collections
* Python

> * dependencies

Reviewer: ```suggestion
* Python and System dependencies 
```

* Python and system dependencies

> * custom user needs

Author: Oh, you're right, great suggestion, i felt that something was wrong with that chapter

which can potentially brake their pipelines.
They might also want to preserve their installation clean from unnecessary packages
and files belonging their colleagues.

Simply put, users might want their content be separated.

Portability
-----------

An Ansible user writes content for Ansible locally and wants to leverage the container technology
to make your automation runtimes portable, shareable and easily deployable to testing and production environments.

What is Execution Environment
=============================

Ansible, as a software application, can run in a container, thus, it can benefit from containerization the same as most other applications.

The Ansible automation execution environment (hereinafter, execution environment or EE) is a container image serving as Ansible control node.

The EE image contains:

* ansible-core
* ansible-runner
* none or more Ansible collections
* Python
* Python and system dependencies
* custom user needs

Tools you can use EEs with
-------------------------

You can use EEs with:

* ansible-navigator
* Ansible AWX
* Automation controller

> * ansible-runner

Reviewer 2: although maybe a bit more advanced, the ansible vscode extention along side the vscode container exention is another nice way to take adavantage of the EE https://code.visualstudio.com/docs/devcontainers/containers

Reviewer 2: (in which case, the shell opens in the container and the tradtional ansible-playboo command can be used)

.. _how_to_prepare_environment:

How to prepare your environment to build and test EEs
=====================================================

Install the following packages:

* podman or docker
* python-pip: to install the tools
* ansible-builder: to build EEs

> * ansible-navigator: to run EEs

Reviewer 2: ansible-navigator depends on ansible-builder which could be noted here

Author: do you think it'd harm if it's not here? I'm trying not to put any info that's not strictly necessary as, generally, users have to install all the mentioned packages in order to build and test EEs anyway

Author: Ah, I see now, yep, makes sense to change, done, thanks!

#Containerization-related applicable things (very briefly, minimum required)

#Ansible control node

#ansible-core

#ansible-collections

Rationale
=========

Software applications typically have dependencies. It can be software libraries, configuration files or other services, to name a few, and Ansible, generally being an exceptional automation tool, is not an exception in terms of the mentioned.

Traditionally, application dependencies are installed by administrator on top of an operating system using packaging systems like RPM or Python-pip.

The major drawback of such approach is that an application might require versions of dependencies different from those provided by default.

In case of Ansible, a typical installation consists of ansible-core and a set of Ansible collections.

At present, there are more than a hundred collections included in the Ansible package and hundreds more are available on Ansible Galaxy and Automation Hub for manual installation. Many of them have dependencies.

The collections can depend on:

* Python versions
* Python packages
* system packages
* other collections

The dependencies have to be installed and sometimes can conflict with each other.

> Moreover, typically an administrator doesn't really need the whole Ansible package with those hundred collections. In most cases, and it's considered a good practice, we need to install only ``ansible-core`` plus a set of collections for specific tasks if needed, no more.

Author: yeah, that's true:) thanks! i follow it but sometimes put the short forms automatically, haven't written the docs for ages:)
